### DISCUSSION TITLE
Add user meta fields to add user/registration form. Set user meta upon user creation.

### DISCUSSION THREAD

**Discussion by gunn4r** [pods-framework/pods#6636] [Ideas]
Looking for a straight forward way I could add custom user meta fields to the add user and/or registration form,  so that I can set the meta data upon user creation instead of having to edit the user after.

**Comment by gunn4r** [comment]
Ran into this issue again so I am going to work on a solution!

**Comment by gunn4r** [comment]
So the necessary hooks needed in the user-new.php file are not in core as of 3.6.x but a patch that adds the hooks has already been created as is slated to be added to 3.7.

See: http://core.trac.wordpress.org/ticket/18709 and http://core.trac.wordpress.org/changeset/25629

**Comment by Shelob9** [comment]
Another user reports wanting to do this, and not being able to here:

http://wordpress.org/support/topic/user-extend-problem

**Comment by sc0ttkclark** [comment]
I think we can revisit this and make it available in some way in Pods 3.1

**Comment by sc0ttkclark** [comment]
Open to other contributors stepping in to have it done sooner, but for now there are other high priority features in my mind that are in 2.7, 2.8, 2.9, and 3.0

**Comment by JoryHogeveen** [comment]
I guess this is for the backend forms right? (wp-admin/***).
I'd say front end forms already support this when configures correctly.

**Comment by yipisaias** [comment]
I've run into this issue and decided to create a plugin to fix it: [pods-custom-user-fields-in-registration.zip](https://github.com/user-attachments/files/17650824/pods-custom-user-fields-in-registration.zip)
Tested with WordPress 6.6.2 and seems to work fine.
I even got it to work properly with required fields.
Some hard-coded strings are in Portuguese though, but I think this is easily adaptable.

**Reply by sinanisler** [reply]
![our-hero-crash](https://github.com/user-attachments/assets/3182e3cd-7280-44f4-8e2b-ab4263e4154c)

but didnt work :D 
"There was a critical error on this site. Please check your site administrator's inbox for instructions. If you are still having problems, please visit"

**Reply by sinanisler** [reply]
ok here this works better not for all fields tho for simple text fields works but not for drop down or relationship..etc


```

// Add custom fields to the new user page
add_action('user_new_form', 'custom_user_fields_add_to_user_new_form', 10, 1);
function custom_user_fields_add_to_user_new_form($operation) {
    if ($operation === 'add-new-user') {
        // Retrieve custom fields for users using Pods
        $user_pod = pods('user');
        $fields = $user_pod->fields();

        echo '<h2>Pods Custom Fields</h2>';
        echo '<table class="form-table">';
        
        foreach ($fields as $field_name => $field_data) {
            // Skip default fields that already exist in the user form
            if (in_array($field_name, ['ID', 'user_login', 'user_pass', 'user_nicename', 'user_email', 'user_url', 'user_registered', 'first_name', 'last_name', 'display_name'])) {
                continue;
            }

            // Check if the field is required in Pods and add the "required" attribute
            $required = isset($field_data['required']) && $field_data['required'] ? 'required' : '';

            // Generate the custom fields
            echo '<tr class="form-field">';
            echo '<th><label for="' . esc_attr($field_name) . '">' . esc_html($field_data['label']);
            if ($required) {
                echo ' <span style="color: red;">*</span>'; // Indicate that the field is required
            }
            echo '</label></th>';
            echo '<td><input type="text" name="pods_meta[' . esc_attr($field_name) . ']" id="' . esc_attr($field_name) . '" class="regular-text" ' . $required . ' /></td>';
            echo '</tr>';
        }

        echo '</table>';
    }
}

// Validate required fields when creating a new user via admin
add_action('user_profile_update_errors', 'custom_user_fields_validate_required_fields', 10, 3);
function custom_user_fields_validate_required_fields($errors, $update, $user) {
    if (!$update && isset($_POST['pods_meta'])) {  // Check if it is a new user creation, not an update
        $user_pod = pods('user');
        
        foreach ($_POST['pods_meta'] as $field_name => $field_value) {
            $field = $user_pod->fields($field_name);

            if (isset($field['required']) && $field['required'] && empty($field_value)) {
                $errors->add('required_field', '<strong>Error:</strong> The field "' . esc_html($field['label']) . '" is required.');
            }
        }
    }
}

// Save custom fields after the user is created
add_action('user_register', 'custom_user_fields_save_custom_user_fields');
function custom_user_fields_save_custom_user_fields($user_id) {
    if (isset($_POST['pods_meta'])) {
        $user_pod = pods('user', $user_id);

        // Save each custom field submitted
        foreach ($_POST['pods_meta'] as $field_name => $field_value) {
            $user_pod->save($field_name, sanitize_text_field($field_value));
        }
    }
}


```
